lib/pipeline_controller.py

```diff
@@ -204,27 +205,32 @@ def _run_active_patrons_single_iteration(self, mode):
             processed_df[['patron_id', 'geoid']] = None
             processed_df[['patron_id', 'geoid']] = processed_df[
                 ['patron_id', 'geoid']].astype('string')
-            processed_df['initial_patron_home_library_code'] = processed_df[
-                'patron_home_library_code']
+            processed_df[
+                ['initial_patron_home_library_code', 'initial_ptype_code']] = \
+                processed_df[['patron_home_library_code', 'ptype_code']]
+            processed_df['is_in_redshift'] = False
+            processed_df['is_in_redshift'] = processed_df['is_in_redshift'].astype('bool')
 
         # For every row not already in Redshift, obfuscate the patron id and
         # geocode it
-        unknown_patrons_df = processed_df[
-            pd.isnull(processed_df['patron_id'])][
-            ['address', 'city', 'region', 'postal_code',
-             'patron_id_plaintext']]
+        unknown_patrons_df = processed_df.loc[
+            ~processed_df['is_in_redshift'],
+            ['address', 'city', 'region', 'postal_code', 'patron_id_plaintext']
+        ]
```

Comment:
Slightly confused by this section, so are you setting everything in the `is_in_redshift` column to False, and then updating it to True with the `unknown_patrons_df` var? What's the reasoning behind this?

Reply:
Yeah it's a little confusing and I was struggling to name my variables clearly, but the idea is that there are three situations and depending on the situation we have to query Redshift either zero, one, or two times for additional info:

1) We're polling for new patrons (mode is NEW_PATRONS and is_in_redshift = False) -- there's no need to query Redshift at all
2a) We're polling for updated patrons and a patron has not changed addresses (mode is UPDATED_PATRONS and is_in_redshift = True) -- we've queried Redshift for the patron's address and found it, so there's no need to geocode the address or get the initial_ptype_code, which was returned by the first query
2b) We're polling for updated patrons and a patrons has changed addresses  (mode is UPDATED_PATRONS and is_in_redshift = False) -- we've queried Redshift for the patron's address and haven't found it, so we have to geocode the new address and then send another query to Redshift to find the patron's initial_ptype_code

Reply:
Gotcha, that makes more sense. This looks good to me then, I'd say at some point maybe adding this info in as a comment or within the README

Reply:
Added to the method description